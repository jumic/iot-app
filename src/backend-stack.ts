import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import { SqsEventSource } from "aws-cdk-lib/aws-lambda-event-sources";
import * as sqs from "aws-cdk-lib/aws-sqs";
import { Construct } from "constructs";
import { ProcessMetricsFunction } from "./lambda/process-metrics-function";

export interface BackendConstructProps {
  table: dynamodb.Table;
  iotDataQueue: sqs.Queue;
}

export class BackendConstruct extends Construct {
  constructor(scope: Construct, id: string, props: BackendConstructProps) {
    super(scope, id);

    const handler = new ProcessMetricsFunction(this, "ProcessMetricsFunction", {
      environment: {
        TABLE_NAME: props.table.tableName,
      },
      memorySize: 2048,
    });
    props.table.grantReadWriteData(handler);

    handler.addEventSource(new SqsEventSource(props.iotDataQueue));
  }
}
