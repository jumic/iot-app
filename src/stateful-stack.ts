import * as dynamodb from '@aws-cdk/aws-dynamodb';
import { Construct, Stack, StackProps } from '@aws-cdk/core';
import * as cdk from '@aws-cdk/core';


export class StatefulStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    // dummy table for deployment test
    new dynamodb.Table(this, 'DummyTable', {
      partitionKey: {
        name: 'id',
        type: dynamodb.AttributeType.STRING,
      },
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
  }
}
