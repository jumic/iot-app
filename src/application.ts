import { Construct, Stage, StageProps } from '@aws-cdk/core';
import { StatefulStack } from './stateful-stack';

export class Application extends Stage {
  constructor(scope: Construct, id: string, props?: StageProps) {
    super(scope, id, props);

    new StatefulStack(this, 'StatefulStack');
  }
}
