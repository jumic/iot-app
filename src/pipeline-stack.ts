import { Construct, Stack, StackProps } from '@aws-cdk/core';
import * as pipelines from '@aws-cdk/pipelines';
import { Application } from './application';

// new MyStack(app, 'my-stack-prod', { env: prodEnv });
export class PipelineStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const pipeline = new pipelines.CodePipeline(this, 'Pipeline', {
      synth: new pipelines.ShellStep('Synth', {
        input: pipelines.CodePipelineSource.connection('jumic/iot-app', 'main', {
          connectionArn: 'arn:aws:codestar-connections:eu-central-1:624132653920:connection/1edb6c36-946f-452d-a051-849b7ffbca7b',
        }),
        commands: [
          'npm ci',
          'npm run build',
          'npx cdk synth',
        ],
      }),
    });

    pipeline.addStage(new Application(this, 'Test', {
      env: {
        account: '632422505464',
        region: 'eu-central-1',
      },
    }));

    pipeline.addStage(new Application(this, 'Prod', {
      env: {
        account: '505825668341',
        region: 'eu-central-1',
      },
    }));
  }
}
