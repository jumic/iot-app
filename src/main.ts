import { App } from '@aws-cdk/core';
import { PipelineStack } from './pipeline-stack';
import { StatefulStack } from './stateful-stack';

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new StatefulStack(app, 'Stateful-stack-dev', { env: devEnv });

new PipelineStack(app, 'IotAppPipelineStack', {
  env: {
    account: '624132653920',
    region: 'eu-central-1',
  },
});

app.synth();
