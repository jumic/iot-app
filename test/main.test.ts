import { Template } from '@aws-cdk/assertions';
import { App } from '@aws-cdk/core';
import { MyStack } from '../src/main';

test('Snapshot', () => {
  const app = new App();
  const stack = new MyStack(app, 'test');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});