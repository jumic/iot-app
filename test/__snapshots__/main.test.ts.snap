// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "BackendConstructProcessMetricsFunctionBC3950FC": Object {
      "DependsOn": Array [
        "BackendConstructProcessMetricsFunctionServiceRoleDefaultPolicyC63930E6",
        "BackendConstructProcessMetricsFunctionServiceRole76F32C27",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "394a297ef26941efae0ef7833643686a314a3a9daf9a33479185ad9fee49118a.zip",
        },
        "Description": "src/lambda/process-metrics.lambda.ts",
        "Environment": Object {
          "Variables": Object {
            "TABLE_NAME": Object {
              "Ref": "StatefulConstructIotTable62D1C116",
            },
          },
        },
        "Handler": "index.handler",
        "MemorySize": 2048,
        "Role": Object {
          "Fn::GetAtt": Array [
            "BackendConstructProcessMetricsFunctionServiceRole76F32C27",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "BackendConstructProcessMetricsFunctionServiceRole76F32C27": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BackendConstructProcessMetricsFunctionServiceRoleDefaultPolicyC63930E6": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "StatefulConstructIotTable62D1C116",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            Object {
              "Action": Array [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "StatefulConstructIotQueue64FEBDA0",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BackendConstructProcessMetricsFunctionServiceRoleDefaultPolicyC63930E6",
        "Roles": Array [
          Object {
            "Ref": "BackendConstructProcessMetricsFunctionServiceRole76F32C27",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "BackendConstructProcessMetricsFunctionSqsEventSourceMyStackStatefulConstructIotQueue5C7ECF0183423864": Object {
      "Properties": Object {
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "StatefulConstructIotQueue64FEBDA0",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "BackendConstructProcessMetricsFunctionBC3950FC",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "StatefulConstructIotQueue64FEBDA0": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "StatefulConstructIotTable62D1C116": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "SK",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "SK",
            "KeyType": "RANGE",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "StatefulConstructTopicRule36E01E2C": Object {
      "Properties": Object {
        "TopicRulePayload": Object {
          "Actions": Array [
            Object {
              "Sqs": Object {
                "QueueUrl": Object {
                  "Ref": "StatefulConstructIotQueue64FEBDA0",
                },
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "StatefulConstructTopicRuleTopicRuleActionRole923198F2",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "AwsIotSqlVersion": "2016-03-23",
          "Sql": "SELECT * as data, topic() as topic, timestamp() as timestamp FROM 'dev/#'",
        },
      },
      "Type": "AWS::IoT::TopicRule",
    },
    "StatefulConstructTopicRuleTopicRuleActionRole923198F2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "iot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StatefulConstructTopicRuleTopicRuleActionRoleDefaultPolicy8F86EF87": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sqs:SendMessage",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "StatefulConstructIotQueue64FEBDA0",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StatefulConstructTopicRuleTopicRuleActionRoleDefaultPolicy8F86EF87",
        "Roles": Array [
          Object {
            "Ref": "StatefulConstructTopicRuleTopicRuleActionRole923198F2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
